Release/1.5.0 (#4516)
aiidateam/aiida-core: CHANGELOG.md

# Changelog

## v1.5.0
(Just for temporary reference, this is the last commit considered: ac4c881ad12cca669d4b7b1745d6dd5f0553a1ba)

> In this minor version release, the support for Python 3.9 is incorporated [[#4301]](https://github.com/aiidateam/aiida-core/pull/4301), while Python 3.5 is dropped [[#4386]](https://github.com/aiidateam/aiida-core/pull/4386).

Reviewer: ```suggestion
In this minor version release, support for Python 3.9 is added [[#4301]](https://github.com/aiidateam/aiida-core/pull/4301), while support for Python 3.5 is dropped [[#4386]](https://github.com/aiidateam/aiida-core/pull/4386).
```

This version is compatible with all current Python versions that are not end-of-life:
 * 3.6
 * 3.7
 * 3.8
 * 3.9

### Features

> - Process functions can now be submitted through the engine [[#4539]](https://github.com/aiidateam/aiida-core/pull/4539)

Reviewer: ```suggestion
- Process functions (`calcfunction` and `workfunction`) can now be submitted to the daemon just like `CalcJob`s and `WorkChain`s [[#4539]](https://github.com/aiidateam/aiida-core/pull/4539)
```

- REST API: list endpoints at base URL [[#4412]](https://github.com/aiidateam/aiida-core/pull/4412)
- `ProcessBuilder`: allow unsetting of inputs through attribute deletion [[#4419]](https://github.com/aiidateam/aiida-core/pull/4419)
- `verdi migrate`: make `--in-place` work across different file systems [[#4393]](https://github.com/aiidateam/aiida-core/pull/4393)

### Improvements
- Added remaining original documentation that didn't make it into the first step of the recent major overhaul of v1.3.0
- Archive importer: refactor the reading code [[#4510]](https://github.com/aiidateam/aiida-core/pull/4510)
- Archive importer: refactor the migrations code [[#4532]](https://github.com/aiidateam/aiida-core/pull/4532)
- Archive exporter: refactor progress bar [[#4504]](https://github.com/aiidateam/aiida-core/pull/4504)

> - Archive exporter: decouple the extraction of data from the writing to file [[#4448]](https://github.com/aiidateam/aiida-core/pull/4448)

Reviewer: Since these concern mostly refactoring and so have merely improvements for developers, should they go to the `Developers` section instead?

Author: Yes, I was debating myself this when I was adding the new ones. I believe at the start we included the first of these here because they also had some performance improvements, but I agree it might just be better to send all of them to `Developers`.

### Bug fixes
- Fix `ZeroDivisionError` in worker slots check [[#4513]](https://github.com/aiidateam/aiida-core/pull/4513)
- `CalcJob`: only attempt to clean up the retrieve temporary folder after parsing if it is present [[#4379]](https://github.com/aiidateam/aiida-core/pull/4379)
- Add missing entry point groups to the mapping [[#4395]](https://github.com/aiidateam/aiida-core/pull/4395)

### Developers
- Replace old format string interpolation with f-strings [[#4400]](https://github.com/aiidateam/aiida-core/pull/4400)
- CI: move `pylint` configuration to `pyproject.toml` [[#4411]](https://github.com/aiidateam/aiida-core/pull/4411)
- CI: use `-e` install for tox + add docker-compose for isolated RabbitMQ [[#4375]](https://github.com/aiidateam/aiida-core/pull/4375)
- CI: add coverage patch threshold to prevent false positives [[#4413]](https://github.com/aiidateam/aiida-core/pull/4413)

> - Archive importer: refactor the reading code [[#4510]](https://github.com/aiidateam/aiida-core/pull/4510)
> - Archive importer: refactor the migrations code [[#4532]](https://github.com/aiidateam/aiida-core/pull/4532)
> - Archive exporter: refactor progress bar [[#4504]](https://github.com/aiidateam/aiida-core/pull/4504)
> - Archive exporter: decouple the extraction of data from the writing to file [[#4448]](https://github.com/aiidateam/aiida-core/pull/4448)

Reviewer 2: ```suggestion
- Archive (import/export) refactor:
     * The refactoring goal was to pave the way for the implementation of a new archive format in v2.0.0 ([ aiidateam/AEP005](https://github.com/aiidateam/AEP/pull/21))
     * Three abstract+concrete interface classes are defined; writer, reader, migrator, which are **independent of the internal structure of the archive**. These classes are used within the export/import code.
     * The code in `aiida/tools/importexport` has been largely re-written, in particular adding `aiida/tools/importexport/archive`, which contains this code for interfacing with an archive, and **does not require connection to an AiiDA profile**.
     * The export logic has been re-written; to minimise required queries (faster), and to allow for "streaming" data into the writer (minimise RAM requirement with new format). It is intended that a similiar PR will be made for the import code.
     * A general progress bar implementation is now available in `aiida/common/progress_reporter.py`. All corresponding CLI commands now also have `--verbosity` option.
     * Merged PRs:
         * Refactor export archive ([#4448](https://github.com/aiidateam/aiida-core/pull/4448) & [#4534](https://github.com/aiidateam/aiida-core/pull/4534))
         * Refactor import archive ([#4510](https://github.com/aiidateam/aiida-core/pull/4510))
         * Refactor migrate archive ([#4532](https://github.com/aiidateam/aiida-core/pull/4532))
         * Add group extras to archive ([#4521](https://github.com/aiidateam/aiida-core/pull/4521))
         * Refactor cmdline progress bar ([#4504](https://github.com/aiidateam/aiida-core/pull/4504) & [#4522](https://github.com/aiidateam/aiida-core/pull/4522))
       * Deprecations: `export_zip`, `export_tar`, `export_tree`, `extract_zip`, `extract_tar` and `extract_tree` functions. `silent` key-word in the `export` function
       * Removed: `ZipFolder` class
```

Author: Since it is a pretty big change with a considerable explanation, I think it would be better to make it a separate section. Check out how it looks now and let me know what you think.

Reviewer 2: yep thats fine ta 👍

- CI: move `pylint` configuration to `pyproject.toml` [[#4411]](https://github.com/aiidateam/aiida-core/pull/4411)
- CI: use `-e` install for tox + add docker-compose for isolated RabbitMQ [[#4375]](https://github.com/aiidateam/aiida-core/pull/4375)
- CI: add coverage patch threshold to prevent false positives [[#4413]](https://github.com/aiidateam/aiida-core/pull/4413)

### Dependencies
- Update requirement `pytest~=6.0` and use `pyproject.toml` [[#4410]](https://github.com/aiidateam/aiida-core/pull/4410)

### Archive (import/export) refactor:
- The refactoring goal was to pave the way for the implementation of a new archive format in v2.0.0 ([ aiidateamAEP005](https://github.com/aiidateam/AEP/pull/21))
- Three abstract+concrete interface classes are defined; writer, reader, migrator, which are **independent of theinternal structure of the archive**. These classes are used within the export/import code.
- The code in `aiida/tools/importexport` has been largely re-written, in particular adding `aiida/toolsimportexport/archive`, which contains this code for interfacing with an archive, and **does not require connectionto an AiiDA profile**.

> - The export logic has been re-written; to minimise required queries (faster), and to allow for "streaming" datainto the writer (minimise RAM requirement with new format). It is intended that a similiar PR will be made for theimport code.

Reviewer 2: typo:

```suggestion
- The export logic has been re-written; to minimise required queries (faster), and to allow for "streaming" datainto the writer (minimise RAM requirement with new format). It is intended that a similiar PR will be made for the import code.
```

### Bug fixes
- `SlurmScheduler`: fix bug in validation of job resources [[#4555]](https://github.com/aiidateam/aiida-core/pull/4555)
- Fix `ZeroDivisionError` in worker slots check [[#4513]](https://github.com/aiidateam/aiida-core/pull/4513)
- `CalcJob`: only attempt to clean up the retrieve temporary folder after parsing if it is present [[#4379]](https://github.com/aiidateam/aiida-core/pull/4379)
- Add missing entry point groups to the mapping [[#4395]](https://github.com/aiidateam/aiida-core/pull/4395)

### Developers
- `verdi group delete`: deprecate and ignore the `--clear` option [[#4357]](https://github.com/aiidateam/aiida-core/pull/4357)
- Replace old format string interpolation with f-strings [[#4400]](https://github.com/aiidateam/aiida-core/pull/4400)
- Improve mypy type checking [[#4553]](https://github.com/aiidateam/aiida-core/pull/4553)
- CI: move `pylint` configuration to `pyproject.toml` [[#4411]](https://github.com/aiidateam/aiida-core/pull/4411)
- CI: use `-e` install for tox + add docker-compose for isolated RabbitMQ [[#4375]](https://github.com/aiidateam/aiida-core/pull/4375)

> - CI: add coverage patch threshold to prevent false positives [[#4413]](https://github.com/aiidateam/aiida-core/pull/4413)

Reviewer 2: ```suggestion
- CI: add coverage patch threshold to prevent false positives [[#4413]](https://github.com/aiidateam/aiida-core/pull/4413)
- CI: Allow for mypy type checking of third-party imports [[#4553]](https://github.com/aiidateam/aiida-core/pull/4553)
```

- The export logic has been re-written; to minimise required queries (faster), and to allow for "streaming" datainto the writer (minimise RAM requirement with new format). It is intended that a similiar PR will be made for the import code.
- A general progress bar implementation is now available in `aiida/common/progress_reporter.py`. All correspondingCLI commands now also have `--verbosity` option.
- Merged PRs:
    - Refactor export archive ([#4448](https://github.com/aiidateam/aiida-core/pull/4448) & [#4534](https://githubcom/aiidateam/aiida-core/pull/4534))
    - Refactor import archive ([#4510](https://github.com/aiidateam/aiida-core/pull/4510))
    - Refactor migrate archive ([#4532](https://github.com/aiidateam/aiida-core/pull/4532))
    - Add group extras to archive ([#4521](https://github.com/aiidateam/aiida-core/pull/4521))
    - Refactor cmdline progress bar ([#4504](https://github.com/aiidateam/aiida-core/pull/4504) & [#4522](https:/github.com/aiidateam/aiida-core/pull/4522))

> - Deprecations: `export_zip`, `export_tar`, `export_tree`, `extract_zip`, `extract_tar` and `extract_tree`functions. `silent` key-word in the `export` function

Reviewer 2: ```suggestion
- Updated archive version from `0.9` -> `0.10`
- Deprecations: `export_zip`, `export_tar`, `export_tree`, `extract_zip`, `extract_tar` and `extract_tree`functions. `silent` key-word in the `export` function
```

# Changelog

## v1.5.0
(Just for temporary reference, this is the last commit considered: 71566ae832ae120037a31591ba0b72305a6f3b2a)

> In this new release of AiiDA the support for python 3.5 is dropped [[#4386]](https://github.com/aiidateam/aiida-core/pull/4386).

Reviewer: ```suggestion
In this minor version release, the support for Python 3.5 is dropped [[#4386]](https://github.com/aiidateam/aiida-core/pull/4386).
```

> This version remains compatible with all current python versions that are not end-of-life:

Reviewer: Python 3.9 is also already out, but we don't support it yet
```suggestion
This version remains compatible with the following current Python versions that are not end-of-life:
```

 * 3.6
 * 3.7
 * 3.8

### Features
- REST API: list endpoints at base URL [[#4412]](https://github.com/aiidateam/aiida-core/pull/4412)
- `ProcessBuilder`: allow unsetting of inputs through attribute deletion [[#4419]](https://github.com/aiidateam/aiida-core/pull/4419)
- `verdi migrate`: make `--in-place` work across different file systems [[#4393]](https://github.com/aiidateam/aiida-core/pull/4393)

### Improvements

> - Refactor Import Archive [[#4510]](https://github.com/aiidateam/aiida-core/pull/4510)

Reviewer: ```suggestion
- Refactor the archive reading code from the import implementation [[#4510]](https://github.com/aiidateam/aiida-core/pull/4510)
```

Author: I changed this based on this recommendation, but I thought it would be better to adapt it to the format of the rest of the archive restructure logs. See first line:

```
- Archive importer: refactor the reading code [[#4510]](https://github.com/aiidateam/aiida-core/pull/4510)
- Archive importer: refactor the migrations code [[#4532]](https://github.com/aiidateam/aiida-core/pull/4532)
- Archive exporter: refactor progress bar [[#4504]](https://github.com/aiidateam/aiida-core/pull/4504)
- Archive exporter: decouple the extraction of data from the writing to file [[#4448]](https://github.com/aiidateam/
```

> - The remaining old content has been incorporated in the new documentation

Reviewer: ```suggestion
- Added remaining original documentation that didn't make it into the first step of the recent major overhaul of v1.3.0
```

- Archive exporter: refactor progress bar [[#4504]](https://github.com/aiidateam/aiida-core/pull/4504)
- Archive exporter: decouple the extraction of data from the writing to file [[#4448]](https://github.com/aiidateam/aiida-core/pull/4448)
- `verdi process show`: order by ctime and print process label [[#4407]](https://github.com/aiidateam/aiida-core/pull/4407)
- `LinkManager`: fix inaccuracy in exception message for non-existent link [[#4388]](https://github.com/aiidateam/aiida-core/pull/4388)

### Bug fixes
- Fix `ZeroDivisionError` in worker slots check [[#4513]](https://github.com/aiidateam/aiida-core/pull/4513)
- Fix `UnboundLocalError` in `aiida.cmdline.utils.edit_multiline_template` [[#4436]](https://github.com/aiidateam/aiida-core/pull/4436)

> - Only remove temporary folder if it is present [[#4379]](https://github.com/aiidateam/aiida-core/pull/4379)

Reviewer: ```suggestion
- `CalcJob`: only attempt to clean up the retrieve temporary folder after parsing if it is present [[#4379]](https://github.com/aiidateam/aiida-core/pull/4379)
```

> - Improve the deprecation warning for `Node.open` outside context manager [[#4434]](https://github.com/aiidateam/aiida-core/pull/4434)

Reviewer: Would make this an improvement

Author: Just to make sure: I see selected all of the entries of bug fixes, you meant all of them? In that case I will directly eliminate the category.

Reviewer: I was talking just about "Improve the deprecation....". We should keep the bug fix category
